kwok/cloudprovider/cloudprovider.go

```diff
@@ -0,0 +1,271 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"sigs.k8s.io/karpenter/pkg/apis/v1beta1"
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+	"sigs.k8s.io/karpenter/pkg/scheduling"
+)
+
+func NewCloudProvider(ctx context.Context, client kubernetes.Interface, instanceTypes []*cloudprovider.InstanceType) *CloudProvider {
+	return &CloudProvider{
+		kubeClient:    client,
+		instanceTypes: instanceTypes,
+	}
+}
+
+type CloudProvider struct {
+	kubeClient    kubernetes.Interface
+	instanceTypes []*cloudprovider.InstanceType
+}
+
+func (c CloudProvider) Create(ctx context.Context, nodeClaim *v1beta1.NodeClaim) (*v1beta1.NodeClaim, error) {
+	// Create the Node because KwoK nodes don't have a kubelet, which is what Karpenter normally relies on to create the node.
+	node, err := c.toNode(nodeClaim)
+	if err != nil {
+		return nil, fmt.Errorf("translating nodeclaim to node, %w", err)
+	}
+	_, err = c.kubeClient.CoreV1().Nodes().Create(ctx, node, metav1.CreateOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("creating node, %w", err)
+	}
+	// convert the node back into a node claim to get the chosen resolved requirement values.
+	return c.toNodeClaim(node)
+}
+
+func (c CloudProvider) Delete(ctx context.Context, nodeClaim *v1beta1.NodeClaim) error {
+	err := c.kubeClient.CoreV1().Nodes().Delete(ctx, nodeClaim.Name, metav1.DeleteOptions{})
+	if err != nil {
+		if kubeErrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("deleting node, %w", err)
+	}
+	return nil
+}
+
+func (c CloudProvider) Get(ctx context.Context, providerID string) (*v1beta1.NodeClaim, error) {
+	nodeName := strings.Replace(providerID, kwokProviderPrefix, "", -1)
+	node, err := c.kubeClient.CoreV1().Nodes().Get(ctx, nodeName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("finding node, %w", err)
+	}
+	return c.toNodeClaim(node)
+}
+
+func (c CloudProvider) List(ctx context.Context) ([]*v1beta1.NodeClaim, error) {
+	nodes, err := c.kubeClient.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("listing nodes, %w", err)
+	}
+	var nodeClaims []*v1beta1.NodeClaim
+	for i, node := range nodes.Items {
+		if !strings.HasPrefix(node.Spec.ProviderID, kwokProviderPrefix) {
+			continue
+		}
+		nc, err := c.toNodeClaim(&nodes.Items[i])
+		if err != nil {
+			return nil, fmt.Errorf("converting machine, %w", err)
+		}
+		nodeClaims = append(nodeClaims, nc)
+	}
+
+	return nodeClaims, nil
+}
+
+// Return the hard-coded instance types.
+func (c CloudProvider) GetInstanceTypes(ctx context.Context, nodePool *v1beta1.NodePool) ([]*cloudprovider.InstanceType, error) {
+	return c.instanceTypes, nil
+}
+
+// Return nothing since there's no cloud provider drift.
+func (c CloudProvider) IsDrifted(ctx context.Context, nodeClaim *v1beta1.NodeClaim) (cloudprovider.DriftReason, error) {
+	return "", nil
+}
+
+func (c CloudProvider) Name() string {
+	return "kwok-provider"
```

Comment:
nit: we just say `aws` in the name for the AWS provider. I think the provider bit is implied. We should check to see what Azure is doing as well